File: web/frontend/src/api/scanner.ts
import api from './client';

// ── Legacy types (kept for backward compatibility) ───────────
export interface OrderCandidate {
  symbol: string;
  direction: string;
  score: number;
  regime: string;
  models_fired: string[];
  entry_price: number;
  stop_loss: number;
  take_profit: number;
  rr_ratio: number;
  approved: boolean;
  rejection_reason: string | null;
  position_size_usdt: number;
  generated_at: string;
}

export interface ScannerResults {
  results: OrderCandidate[];
  count: number;
  scanner_running: boolean;
}

export interface WatchlistResponse {
  symbols: string[];
  weights: Record<string, number>;
}

// ── Phase 3B: Pipeline Status types ──────────────────────────
export interface PipelineDiagnostics {
  candle_count: number;
  candle_age_s: number;
  candle_ts_str: string;
  regime_confidence: number;
  regime_probs: Record<string, number>;
  all_model_names: string[];
  models_disabled: string[];
  models_fired: string[];
  models_no_signal: string[];
  signal_details: Record<string, { direction: string; strength: number }>;
  indicator_cols_missing: string[];
  pre_filter_reason: string;
  rejection_reason: string;
  // ── MIL Phase 4A diagnostic keys (pass-through) ──────────
  mil_active?: boolean;
  mil_influence_pct?: number;
  mil_total_delta?: number;
  mil_capped?: boolean;
  mil_dominant_source?: string;
  mil_breakdown?: MILBreakdown;
  mil_funding_signal?: number;
  mil_funding_percentile?: number;
  mil_funding_divergence?: boolean;
  mil_funding_weighted_rate?: number;
  mil_oi_delta?: number;
  mil_oi_delta_4h?: number;
  mil_liquidation_proximity?: number;
  mil_oi_volume_ratio?: number;
  mil_orchestrator_meta?: number;
  mil_veto_active?: boolean;
}

// ── Phase S1: MIL breakdown (with Phase 4B placeholders) ────
export interface MILBreakdown {
  orchestrator_delta: number;
  sentiment_delta: number;       // Phase 4B placeholder (0.0 until wired)
  news_delta: number;            // Phase 4B placeholder (0.0 until wired)
  other_orchestrator_delta: number; // orchestrator minus sentiment/news attribution
  oi_delta: number;
  liquidation_delta: number;
}

export type PipelineStatus =
  | 'Eligible'
  | 'Risk Blocked'
  | 'No Signal'
  | 'Regime Filtered'
  | 'Pre-Filter'
  | 'Waiting'
  | 'Error';

export interface PipelineRow {
  asset_id: number;
  symbol: string;
  allocation_weight: number;
  price: number | null;
  regime: string;
  regime_confidence: number;
  models_fired: string[];
  models_no_signal: string[];
  score: number;
  direction: string;
  status: PipelineStatus;
  reason: string;
  is_approved: boolean;
  entry_price: number | null;
  stop_loss: number;
  take_profit: number;
  rr_ratio: number;
  position_size_usdt: number;
  scanned_at: string;
  // ── Phase S1: promoted MIL fields (top-level) ──────────────
  technical_score: number;
  final_score: number;
  mil_active: boolean;
  mil_total_delta: number;
  mil_influence_pct: number;
  mil_capped: boolean;
  mil_dominant_source: string;
  mil_breakdown: MILBreakdown | Record<string, never>;
  // ── Phase S1: decision explainability ──────────────────────
  decision_explanation: string;
  block_reasons: string[];
  // ── diagnostics (unchanged) ────────────────────────────────
  diagnostics: PipelineDiagnostics | Record<string, never>;
}

export interface PipelineSummary {
  total: number;
  eligible: number;
  active_signals: number;
  blocked: number;
}

export interface PipelineStatusResponse {
  status: string;
  pipeline: PipelineRow[];
  summary: PipelineSummary;
  scanner_running: boolean;
  last_scan_at: string;
  source: string;
}

// ── API functions ────────────────────────────────────────────
export async function getScannerResults(): Promise<ScannerResults> {
  const resp = await api.get<ScannerResults>('/scanner/results');
  return resp.data;
}

export async function getPipelineStatus(): Promise<PipelineStatusResponse> {
  const resp = await api.get<PipelineStatusResponse>('/scanner/pipeline-status');
  return resp.data;
}

export async function getWatchlist(): Promise<WatchlistResponse> {
  const resp = await api.get<WatchlistResponse>('/scanner/watchlist');
  return resp.data;
}

export async function triggerScan(): Promise<{ status: string; message: string }> {
  const resp = await api.post('/scanner/trigger');
  return resp.data;
}

File: web/backend/app/api/scanner.py
# ============================================================
# NEXUS TRADER Web — Scanner Router
#
# GET  /scanner/results          — last scan cycle results (approved only)
# GET  /scanner/watchlist        — current watchlist + weights
# GET  /scanner/pipeline-status  — Phase 3B: full per-asset pipeline dashboard
# POST /scanner/trigger          — trigger immediate scan
# ============================================================
from __future__ import annotations

import logging

from fastapi import APIRouter, Depends

from app.auth.dependencies import get_current_user
from app.api.engine import _send_engine_command

logger = logging.getLogger(__name__)
router = APIRouter(
    prefix="/scanner",
    tags=["scanner"],
    dependencies=[Depends(get_current_user)],
)


@router.get("/results")
async def scanner_results():
    """Return results from the most recent scan cycle."""
    return await _send_engine_command("get_scanner_results", {})


@router.get("/pipeline-status")
async def pipeline_status():
    """Return full per-asset pipeline status for all tradable assets.

    Phase 3B: Shows every tradable asset from Asset Management and where
    it stands in the scan → regime → strategy → confluence → risk pipeline.
    Assets that have not been scanned yet appear with status 'Waiting'.

    Response schema:
        {
            "status": "ok",
            "pipeline": [
                {
                    "asset_id": int,
                    "symbol": str,
                    "allocation_weight": float,
                    "price": float | null,
                    "regime": str,
                    "regime_confidence": float,
                    "models_fired": list[str],
                    "models_no_signal": list[str],
                    "score": float,
                    "direction": str,
                    "status": str,  // Eligible | Risk Blocked | No Signal | Regime Filtered | Pre-Filter | Waiting | Error
                    "reason": str,
                    "is_approved": bool,
                    "entry_price": float | null,
                    "stop_loss": float,
                    "take_profit": float,
                    "rr_ratio": float,
                    "position_size_usdt": float,
                    "scanned_at": str,
                    "diagnostics": dict,
                }
            ],
            "summary": {
                "total": int,
                "eligible": int,
                "active_signals": int,
                "blocked": int,
            },
            "scanner_running": bool,
            "last_scan_at": str,
            "source": str,
        }
    """
    return await _send_engine_command("get_pipeline_status", {})


@router.get("/watchlist")
async def watchlist():
    """Return current scanner watchlist with symbol weights."""
    return await _send_engine_command("get_watchlist", {})


@router.post("/trigger")
async def trigger_scan():
    """Trigger an immediate scan cycle."""
    return await _send_engine_command("trigger_scan", {})
